__tests__/integration/query-typings.test.ts

```diff
@@ -27,4 +28,28 @@ describe.each`
       }
     }
   });
+
+  it("allows customers to infer their own types in queries from fql statements", async () => {
+    // This is a noop function that is only used to validate the inferred type of the query
+    // It will fail at build time if types are not inferred correctly.
+    const noopToValidateInferredType = (value: MyType) => {};
+
+    const query = fql<MyType>`{ "x": 123 }`;
+    const q2 = fql`{ "x": ${query} }`;
+
+    expect.assertions(1);
+    if ("query" === method) {
+      const result = (await client.query(query)).data;
+      noopToValidateInferredType(result);
+      expect(result).toEqual({ x: 123 });
+    } else {
+      for await (const page of client.paginate<MyType>(fql`{ "x": 123}`)) {
```

Comment:
q2 is unused. 

Also, while providing a query that returns a single value is technically allowed, it's not "proper". It's really just a fallback which takes the return value and turns it into an array with a single item: the result of the query.

We should test a query (and type it so) that returns the FQL type `Set`. We should do this in the above test as well.

If we do the following, we actually, currently get a type error, because the current signature of `Client.paginate` accepts a `Query<T>` as argument, but what we (typically) want is the argument to be `Query<Page<T>>`. I guess we need to allow both.

```js
    const query = fql<MyType>`{ "x": 123 }`;
    const paginatedQuery = fql`[{ "x": 123 }].toSet()`;

    expect.assertions(1);
    if ("query" === method) {
      const result = (await client.query(query)).data;
      noopToValidateInferredType(result);
      expect(result).toEqual({ x: 123 });
    } else {
      // type error if `Query<Page<T>>` is not allowed as an argument
      for await (const page of client.paginate<MyType>(paginatedQuery) { 
     
  // ...
```